File: dsa-problems/LinkedLists/solutions/cpp/LNK-001-lab-roster-append.cpp
#include <iostream>
#include <sstream>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

class Solution {
public:
    ListNode* head = nullptr;
    ListNode* tail = nullptr;

    void pushBack(int value) {
        ListNode* newNode = new ListNode(value);
        if (!head) {
            head = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }
    }

    vector<int> toArray() {
        vector<int> result;
        ListNode* current = head;
        while (current) {
            result.push_back(current->val);
            current = current->next;
        }
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    string dummy;
    getline(cin, dummy);

    Solution solution;
    for (int i = 0; i < n; i++) {
        string line;
        getline(cin, line);
        if (line.rfind("push_back", 0) == 0) {
            stringstream ss(line);
            string op;
            int value;
            ss >> op >> value;
            solution.pushBack(value);
        } else {
            vector<int> arr = solution.toArray();
            for (int j = 0; j < (int)arr.size(); j++) {
                if (j) cout << " ";
                cout << arr[j];
            }
            cout << "\n";
        }
    }
    return 0;
}

File: dsa-problems/LinkedLists/solutions/java/LNK-001-lab-roster-append.java
import java.util.*;

class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

class Solution {
    private ListNode head = null;
    private ListNode tail = null;

    public void pushBack(int value) {
        ListNode newNode = new ListNode(value);
        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
    }

    public int[] toArray() {
        List<Integer> list = new ArrayList<>();
        ListNode current = head;
        while (current != null) {
            list.add(current.val);
            current = current.next;
        }
        // Convert List to int[]
        int[] arr = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            arr[i] = list.get(i);
        }
        return arr;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        sc.nextLine();

        Solution solution = new Solution();
        for (int i = 0; i < n; i++) {
            String line = sc.nextLine().trim();
            if (line.startsWith("push_back")) {
                String[] parts = line.split(" ");
                int value = Integer.parseInt(parts[1]);
                solution.pushBack(value);
            } else {
                int[] arr = solution.toArray();
                for (int j = 0; j < arr.length; j++) {
                    System.out.print(arr[j]);
                    if (j + 1 < arr.length) System.out.print(" ");
                }
                System.out.println();
            }
        }
        sc.close();
    }
}
